source/common/grpc/async_client_impl.cc

```diff
@@ -234,7 +242,8 @@ void AsyncStreamImpl::onData(Buffer::Instance& data, bool end_stream) {
       streamError(Status::WellKnownGrpcStatus::Internal);
       return;
     }
-    if (!callbacks_.onReceiveMessageRaw(frame.data_ ? std::move(frame.data_)
+    if (!waiting_to_delete_on_remote_close_ &&
```

Comment:
I think this worth some discussion. 
imagine that the grpc-client sends a request to grpc server, and immediately half-close the stream, it's still expected to receive and process the response from grpc server. 

https://github.com/envoyproxy/envoy/blob/main/envoy/grpc/async_client.h#L55-L60

Reply:
You are confusing client half-close and the step for waiting for the server to gracefully close the stream to avoid reset. Client half-closing the stream does not mean it can not receive anything from the server, it still can.

Reply:
I see it now.
IIUC besides closeStream and resetStream, you are introducing a 3rd method waitForRemoteCloseAndDelete that's supposed to be called at filter destruction time only? 
do you plan to extend usage of this method to other fileters that involves side grpc calls? 

OTOH, if we goal is to do "half-close whenever feasible" from ext_proc, if we ever have that, this method would be less useful?
would https://github.com/envoyproxy/envoy/pull/37083 be an option too, which doesn't introduce a new "closing method" in the grpc-lib.

Reply:
This is the first place for now. I'm going to test it internally and then flip the runtime flag to true. Then we can look at other filters - I think there is only one, RLQS, which is using streaming gRPC.

This method will always needed. There are legitimate cases in ext_proc where it will be called. For example when server sends instant reply command. Also observability mode will use it as well. It separates lifetimes of the main stream and side stream.

Wasn't #37083 causing crashes?

Reply:
I think we never identified whether #37083 causes crashes or not. 

Sounds good if you think this method is generic enough to be used elsewhere. 
my main point is if ext_proc or any other streaming grpc client user could do half-close and then reset upon receiving trailers from remote ext-proc grpc server, we don't need this 3rd state.

Reply:
Not many clients will wait for trailers if their state machine has completed, since it can add unnecessary latency.